CIPs/CIP-124/cip-124.md

```diff
@@ -0,0 +1,321 @@
+<EventID> ::= <multibase-prefix><multicodec-streamid><stream-type>
+  <genesis-cid-bytes>
+  <multicodec-unsigned-varint><event hight>
+  <event-cid-bytes>
+```
+[multiformats/unsigned-varint](https://github.com/multiformats/unsigned-varint)
+
+#### **Stream set sort order**
+example lexicographical sort order
+```
+ceramic/<network>/<partitions_key>/<EventID>
+```
+
+We could partition on `model`, `controller`, `time` , or something else using a partition key? We could even map multiple dimensions into a single partition key and then partition based on the calulated partition key.
+
+What are the sorting orders that make sense?
+- Total time ordering of all events in all streams.
+    - `ceramic/<network>/<timeID>/<EventID>`
+- Sort by model/StreamID
+    - `ceramic/<network>/<model>/<EventID>`
+- Sort by controller/model/StreamID
+    - `ceramic/<network>/<did>/<model>/<EventID>`
+- Sort by StreamID lexicographic
+    - `ceramic/<network>/<EventID>`
+- Time / block height, causal or lexicographic within stream
+    - `ceramic/<network>/<StreamID>/<block height><EventID>`
+- Z-0rder (Hilbert Curve order)
+    - `ceramic/<network>/< z(model, controller, time) >`
+    - [https://youtu.be/YLVkITvF6KU](https://youtu.be/YLVkITvF6KU)
+    - [https://en.wikipedia.org/wiki/Z-order](https://en.wikipedia.org/wiki/Z-order)
+
+### **Interest Advertisements**
+In order to discover peers with overlaping interests nodes must avertise there interest ranges to the network.
+
+#### **DHT peer discovery**
+1. If known peer list too small
+    - Find providers in the DHT for the given StreamSetID `ceramic/<network>/<partition_key>/` (e.g. model)
+    - Add the node to your known peer list if query receves a reponce.
+    - Advertise your node's stream set ranges of interests on the DHT.
+2. Send query to provider(s)
+    - Select peer(s) at random
+    - Maintain an internal reputation list for peers
+        - How up to date are they.
+        - Latency.
+        - Interests
+    - The query should include data about:
+        - The stream events you want.
+        - The stream events you have.
+        - If out of sync recur.
+3. Providers responds with their interests
+    - The response should include data about:
+        - The stream events they have.
+        - The stream events they want.
+        - If out of sync recur.
+    - The node should add you to their known peer list after responding.
+    
+
+Periodically repeat the process above to achieve eventual consistency.
+
+```ts
+type InterestRange struct {
+	start EventID // A string less the any event of intrest
+	dividers [
+    {
+      associative_hash string // the associative hash of all EventIDs in sub-range
+      link CID // [optional] the CID of the subnode
+
+      // an EventID larger then any EventID in the preceding sub-range
+      // and less then any in the subsuquent sub-range
+      stop string 
+
+    }
+  ]
+}
+
+type Interests: InterestRange[]
+```
+
+#### **Interactive**
```

Comment:
Is this section supposed to describe the interactive algorithm between nodes? 
I like the use of animal names, but you should clarify that this is used instead of the long identifiers that were defined above.
Also feel like this section could be cleaned up a bit more to be easier to follow.